tiledb/sm/serialization/serializable_stats.cc

```diff
@@ -0,0 +1,132 @@
+ */
+
+// clang-format off
+#ifdef TILEDB_SERIALIZATION
+#include <capnp/compat/json.h>
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#include "tiledb/sm/serialization/capnp_utils.h"
+#include "serializable_stats.h"
+#endif
+// clang-format on
+
+#include "tiledb/common/common.h"
+
+namespace tiledb {
+namespace sm {
+namespace serialization {
+
+#ifdef TILEDB_SERIALIZATION
+
+void stats_from_capnp(
+    const capnp::Stats::Reader& stats_reader, stats::Stats* stats) {
+  if (stats_reader.hasCounters()) {
+    auto counters = stats->counters();
+    auto counters_reader = stats_reader.getCounters();
+    for (const auto entry : counters_reader.getEntries()) {
+      (*counters)[std::string(entry.getKey().cStr())] = entry.getValue();
+    }
+  }
+
+  if (stats_reader.hasTimers()) {
+    auto timers = stats->timers();
+    auto timers_reader = stats_reader.getTimers();
+    for (const auto entry : timers_reader.getEntries()) {
+      (*timers)[std::string(entry.getKey().cStr())] = entry.getValue();
+    }
+  }
+}
+
+std::string SerializableStats::to_json() {
+  ::capnp::MallocMessageBuilder message;
+  auto builder = message.initRoot<capnp::Stats>();
+
+  to_capnp(builder);
+  ::capnp::JsonCodec json;
+  kj::String capnp_json = json.encode(builder);
+
+  return std::string(capnp_json.cStr(), capnp_json.size());
+}
+
+void SerializableStats::to_capnp(capnp::Stats::Builder& stats_builder) {
+  // Build counters
+  if (!counters_.empty()) {
+    auto counters_builder = stats_builder.initCounters();
+    auto entries_builder = counters_builder.initEntries(counters_.size());
+    uint64_t index = 0;
+    for (const auto& entry : counters_) {
+      entries_builder[index].setKey(entry.first);
+      entries_builder[index].setValue(entry.second);
+      ++index;
+    }
+  }
+
+  // Build timers
+  if (!timers_.empty()) {
+    auto timers_builder = stats_builder.initTimers();
+    auto entries_builder = timers_builder.initEntries(timers_.size());
+    uint64_t index = 0;
+    for (const auto& entry : timers_) {
+      entries_builder[index].setKey(entry.first);
+      entries_builder[index].setValue(entry.second);
+      ++index;
+    }
+  }
+}
+
+stats::Stats* SerializableStats::from_capnp(
+    stats::Stats* parent_stats, const capnp::Stats::Reader& stats_reader) {
+  std::unordered_map<std::string, uint64_t> counters;
+  if (stats_reader.hasCounters()) {
+    auto counters_reader = stats_reader.getCounters();
+    for (const auto entry : counters_reader.getEntries()) {
+      counters[std::string(entry.getKey().cStr())] = entry.getValue();
+    }
+  }
+
+  std::unordered_map<std::string, double> timers;
+  if (stats_reader.hasTimers()) {
+    auto timers_reader = stats_reader.getTimers();
+    for (const auto entry : timers_reader.getEntries()) {
+      timers[std::string(entry.getKey().cStr())] = entry.getValue();
+    }
+  }
+
+  return parent_stats->create_child("Subarray", move(timers), move(counters));
```

Comment:
How would a `SerializableStats` object know it's going to serve a `Subarray`? 
I would ditch the `stats_from_capnp` leftover function altogether and 1) make `SerializableStats::from_capnp` C41 construct a `Stats` object (no name, name is a property of a Stats position in the hierarchy anyway)
2) add `Stats::init_from(Stats&& stats)` which we call like `stats->init_from(SerializableStats::from_capnp(args))`
Let the future `Stats` redesign handle the ugliness above, but allow us to have a clean start with this serialization refactoring.